### PR TITLE
#37 2Split upsert into create and update into repository

### PR DESCRIPTION
**Pull request by alesyalis**
Create GetCandidateWithProjectAsync, UpsertFeedbackAsync into candidate Controller
create ValidationException
сreate NotFoundException
create GetCandidateWithProjectAsync, UpsertFeedbackAsync into CandidateService
create CreateFeedbackAsync,CreateFeedbackAsync,  UpdateWithArrayFiltersAsync into CandidateRepository

add try catch

### CODE REVIEW

File: Recrutify/Recrutify.DataAccess/Models/CandidateStatusFeedBack.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Recrutify.DataAccess.Models
+{
+    public class CandidateStatusFeedBack
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/CandidateStatusFeedBack.cs:6]: remove class

File: Recrutify/Recrutify.DataAccess/Recrutify.DataAccess.csproj [hunk 2]
```diff
@@ -17,4 +17,8 @@
     </PackageReference>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Models\NewFolder\" />
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Recrutify.DataAccess.csproj:21]: remove

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs [hunk 3]
```diff
@@ -35,20 +33,28 @@ public async Task UpsertFeedbackAsync(Guid id, Guid projectId, Feedback feedback
...
 
-            var updateResult = await GetCollection().UpdateOneAsync(filter, updateDefinition, updateOptions);
-            if (updateResult.ModifiedCount == 0)
-            {
-                var projectResultArrayFilters = new List<ArrayFilterDefinition>
+            return UpdateWithArrayFiltersAsync(id, updateDefinition, arrayFilters);
+        }
+
+        public Task CreateFeedbackAsync(Guid id, Guid projectId, Feedback feedback)
+        {
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:41]: use same order of vars as in Update: updateBuilder, updateDefinition, binaryProjectId

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs [hunk 4]
```diff
@@ -19,23 +20,23 @@ public CandidateProfile()
...
             CreateMap<StatusDTO, Status>();
 
             CreateMap<FeedbackType, FeedbackTypeDTO>()
-                .ConvertUsing(x => (FeedbackTypeDTO)((int)x));
+                 .ConvertUsing(x => (FeedbackTypeDTO)((int)x));
             CreateMap<FeedbackTypeDTO, FeedbackType>();
 
             CreateMap<EnglishLevel, EnglishLevelDTO>()
-                .ConvertUsing(x => (EnglishLevelDTO)((int)x));
+                 .ConvertUsing(x => (EnglishLevelDTO)((int)x));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:39]: remov extra spaces lines 31, 35, 39

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs [hunk 5]
```diff
@@ -37,19 +39,39 @@ public async Task<ActionResult<CandidateDTO>> CreateAsync(CandidateCreateDTO can
...
+            {
+                return BadRequest(ex.Errors.FirstOrDefault());
+            }
 
-            await _candidateService.UpsertFeedbackAsync(id, projectId, feedbackDto);
             return NoContent();
         }
 
+        [HttpGet("{id:guid}/{projectId:guid}")]
+        public async Task<ActionResult<CandidateDTO>> UpsertAllFeedbackAsync(Guid id, Guid projectId)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:62]: GetCandidateWithProjectAsync

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 6]
```diff
@@ -46,10 +52,51 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
...
+            }
+
+            var currentProjectResult = candidate.ProjectResults?.FirstOrDefault(x => x.ProjectId == projectId);
+            candidate.ProjectResults = new List<ProjectResult> { currentProjectResult };
+            return _mapper.Map<CandidateDTO>(candidate);
+        }
+
+        public async Task UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
+        {
+            var projectResultWithFeedback = await _candidateRepository.GetAsync(id);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:70]: candidate

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 7]
```diff
@@ -46,10 +52,51 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
...
+                ?.Feedbacks?.FirstOrDefault(x => x.UserId == feedbackDto.UserId && x.Type == _mapper.Map<FeedbackType>(feedbackDto.Type));
+            if (currentFeedback == null)
+            {
+                var newFeedback = _mapper.Map<Feedback>(feedbackDto);
+                newFeedback.CreatedOn = DateTime.UtcNow;
+                await _candidateRepository.CreateFeedbackAsync(id, projectId, newFeedback);
+            }
+            else
+            {
+                var projectresult = projectResultWithFeedback.ProjectResults.FirstOrDefault(x => x.ProjectId == projectId);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:86]: rename to projectResult, move it to line 75, and on line 76 write projectResult?.Feedbacks?.FirstOrDefault(x => x.UserId == feedbackDto.UserId && x.Type == _mapper.Map<FeedbackType>(feedbackDto.Type));

Now you do search by projectId twice - on 76, 86 lines, it's bad

File: Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs [hunk 8]
```diff
@@ -1,4 +1,5 @@
-﻿using FluentValidation;
+﻿using System;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs:1]: this using necessary?

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+                .WithMessage("Сannot be updated from one day after creation");
+            RuleFor(f => f.Status)
+                .IsInEnum()
+                .Must(s => !new[] { Status.Accepted, Status.Declined, Status.WaitingList, }.Contains(s))
+                .WithMessage("Cannot be updated for candidate in current status");
+        }
+
+        private bool ShouldBeNoMoreAValidTime(Feedback date)
+        {
+            var feedbackCreationDate = date.CreatedOn;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:24]: you can remove this var and write directly date.CreatedOn

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+            var result = DateTime.Now.Day - feedbackCreationDate.Day;
+            return result <= 1;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:26]: can write (DateTime.Now.Day - date.CreatedOn.Day) <=1

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+                .NotNull()
+                .Must(ShouldBeNoMoreAValidTime)
+                .WithMessage("Сannot be updated from one day after creation");
+            RuleFor(f => f.Status)
+                .IsInEnum()
+                .Must(s => !new[] { Status.Accepted, Status.Declined, Status.WaitingList, }.Contains(s))
+                .WithMessage("Cannot be updated for candidate in current status");
+        }
+
+        private bool ShouldBeNoMoreAValidTime(Feedback date)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs:22]: you can refactor method with recommendations below, remove this method and write it directly in Must like with Status rule.

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs [hunk 12]
```diff
@@ -37,19 +39,39 @@ public async Task<ActionResult<CandidateDTO>> CreateAsync(CandidateCreateDTO can
         }
 
         // [Authorize(Policy = Constants.Constants.Policies.FeedbackPolicy)]
-        [HttpPut]
-        public async Task<ActionResult> UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
+        [HttpPut("feedback")]
+        public async Task<ActionResult<CandidateDTO>> UpsertFeedbackAsync(Guid id, Guid projectId, CreateFeedbackDTO feedbackDto)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:43]: just Task<ActionResult>

Comment (elizavetapavliv) [reply Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:43]: without CandidateDTO

File: Recrutify/Recrutify.DataAccess/Models/CandidateStatusFeedBack.cs (hunk 1 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Models/CandidateStatusFeedBack.cs:6]: done

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs (hunk 12 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:43]: done

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 10 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs]: done

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 9 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs]: done

File: Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs (hunk 8 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/UpdateProjectValidator.cs]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 7 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: donedo

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 6 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: done

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs (hunk 4 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs]: done

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs (hunk 5 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Controllers/CandidatesController.cs]: done

File: Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs (hunk 3 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Repositories/CandidateRepository.cs:41]: done

File: Recrutify/Recrutify.DataAccess/Recrutify.DataAccess.csproj (hunk 2 above)
Comment (alesyalis) [reply Recrutify/Recrutify.DataAccess/Recrutify.DataAccess.csproj]: done

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs [hunk 13]
```diff
@@ -18,24 +19,24 @@ public CandidateProfile()
...
+                .ForMember(dest => dest.CreatedOn, opt => opt.Ignore());
             CreateMap<Feedback, FeedbackDTO>();
 
             CreateMap<ContactDTO, Contact>().ReverseMap();
 
             CreateMap<LocationDTO, Location>().ReverseMap();
 
             CreateMap<Status, StatusDTO>()
-                .ConvertUsing(x => (StatusDTO)((int)x));
+                 .ConvertUsing(x => (StatusDTO)(int)x);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:31]: spaces before dots remove

File: Recrutify/Recrutify.Services/Exceptions/ValidationException.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
+﻿using Microsoft.AspNetCore.Mvc.ModelBinding;
+using System;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Exceptions/ValidationException.cs:2]: fix warnings

File: Recrutify/Recrutify.Services/Validators/UpsertFeedbackValidator.cs [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+            RuleFor(f => f.TextFeedback)
+                .NotNull()
+                .NotEmpty()
+                .MaximumLength(500);
+            RuleFor(f => f.Rating)
+                .NotEmpty();
+            RuleFor(f => f.UserId)
+                .NotEmpty();
+            RuleFor(f => f.Type)
+                .NotNull()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/UpsertFeedbackValidator.cs:19]: cannot be null and empty remove

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs (hunk 13 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs]: done

File: Recrutify/Recrutify.Services/Exceptions/ValidationException.cs [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+
+namespace Recrutify.Services.Exceptions
+{
+    public class ValidationException : Exception
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Exceptions/ValidationException.cs:7]: remove this class

Comment (alesyalis) [reply Recrutify/Recrutify.Services/Exceptions/ValidationException.cs:7]: done

File: Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs (hunk 11 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/ProjectResultValidator.cs]: done

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**